ipwb/replay.py

```diff
@@ -179,6 +180,111 @@ def showMementosForURIRs_sansJS():
     return redirect('/memento/*/' + urir, code=301)
 
 
+def bin_search(iter, key, datetime=None):
+    # Read line encompassing current position
+    ln = iter.readline()
+
+    while ipwbUtils.isCDXJMetadataRecord(ln):
+        ln = iter.readline()
+
+    surtk, datetimeK, rest = ln.split(maxsplit=2)
+    datetimeK = datetimeK.decode()
+
+    matchDegree = getMatchDegree(surt, datetime, surtk, datetimeK)
+    if matchDegree == MementoMatch.EXACTMATCH:
+        return [ln]
+
+    # If further searching required...
+    left = 0
+    # Go to end of seek stream
+    iter.seek(0, 2)
+    right = iter.tell()  # Current position
+
+    lines = set()  # Prevents dupes
+    while (right - left > 1):
+        mid = (right + left) // 2
+        iter.seek(mid)
+        iter.readline()  # Purge rest of current line
+        line = iter.readline()  # Read the next full line
+
+        if len(line) == 0:
+            right = mid
+            continue
+
+        surtk, datetimeK, rest = line.split(maxsplit=2)
+        datetimeK = datetimeK.decode()
+
+        surtk = surtk.rstrip(b"/")
+        key = key.rstrip(b"/")
+
+        matchDegree = getMatchDegree(key, datetime, surtk, datetimeK)
+
+        if matchDegree == MementoMatch.RIGHTKEYWRONGDATE:
+            lines.add(line)
+            # Iterate further to get lines after selection point
+            nextLine = iter.readline()
+            while nextLine:
+                surtk, datetimeK, rest = nextLine.split(maxsplit=2)
+                surtk = surtk.rstrip(b"/")
+                datetimeK = datetimeK.decode()
+
+                matchDegree = getMatchDegree(key, datetime, surtk, datetimeK)
+                if matchDegree == MementoMatch.RIGHTKEYWRONGDATE:
+                    lines.add(nextLine)
+                elif matchDegree == MementoMatch.EXACTMATCH:
+                    # Exact match found while iterating
+                    lines.clear()
+                    lines.add(nextLine)
+                    break
```

Comment:
I think the logic here needs to be reworked and changed from `break` to `return` as the outer loop will continue to execute, as a result, despite clearing any prior non-exact matches, more non-exact matches might be added to the set that follow the exact match.

```suggestion
                elif matchDegree == MementoMatch.EXACTMATCH:
                    # Exact match found while iterating
                    return [nextLine]
```

Reply:
I agree. The programmatic flow changed here with regard to manipulating the progressively built `set()`. Change applied in 854af87.